### PR TITLE
Add networksettings to reward providers

### PR DESCRIPTION
**Pull request by andrewcoh**
### Proposed change(s)

Describe the changes made in this PR.

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: config/imitation/Hallway.yaml [hunk 1]
```diff
@@ -25,7 +25,6 @@ behaviors:
       gail:
         gamma: 0.99
         strength: 0.01
-        encoding_size: 128
```
Comment (andrewcoh) [inline config/imitation/Hallway.yaml:28]: using default network settings

File: ml-agents/mlagents/trainers/settings.py [hunk 2]
```diff
@@ -199,28 +199,37 @@ def structure(d: Mapping, t: type) -> Any:
             enum_key = RewardSignalType(key)
             t = enum_key.to_settings()
             d_final[enum_key] = strict_to_cls(val, t)
+            if "encoding_size" in val:
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/settings.py:202]: Backward compatible with old configs

File: ml-agents/mlagents/trainers/settings.py [hunk 3]
```diff
@@ -199,28 +199,37 @@ def structure(d: Mapping, t: type) -> Any:
...
+                        "encoding_size"
+                    ]
         return d_final
 
 
 @attr.s(auto_attribs=True)
 class GAILSettings(RewardSignalSettings):
-    encoding_size: int = 64
     learning_rate: float = 3e-4
+    encoding_size: int = 0
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/settings.py:217]: need to keep so cattrs doesnt error if its specified.

Comment (ervteng) [reply ml-agents/mlagents/trainers/settings.py:217]: Will it break if it's Optional[int] = None?

File: ml-agents/mlagents/trainers/torch/components/reward_providers/gail_reward_provider.py [hunk 4]
```diff
@@ -75,13 +74,8 @@ def __init__(self, specs: BehaviorSpec, settings: GAILSettings) -> None:
...
 
-        encoder_settings = NetworkSettings(
-            normalize=settings.normalize,
-            hidden_units=settings.encoding_size,
-            num_layers=2,
-            vis_encode_type=EncoderType.SIMPLE,
-            memory=None,
-        )
+        encoder_settings = settings.network_settings
+        print(encoder_settings.hidden_units)
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/torch/components/reward_providers/gail_reward_provider.py]: Delete the print

**Comment by ervteng** [comment]
cc @hvpeteet, @sini - this will introduce a (backwards-compatible) change for the YAMLs - just checking to make sure it won't cause any issues.

File: ml-agents/mlagents/trainers/settings.py [hunk 5]
```diff
@@ -183,7 +183,7 @@ def to_settings(self) -> type:
 class RewardSignalSettings:
     gamma: float = 0.99
     strength: float = 1.0
-    normalize: bool = False
+    network_settings: NetworkSettings = attr.ib(factory=NetworkSettings)
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/settings.py:186]: I would make this one optional and if it is None, then use the Policy's network settings rather than our own defaults. How does that sound ?

File: ml-agents/mlagents/trainers/settings.py (hunk 2 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/settings.py:202]: Can you ad a comment around this code so we will remember it ?

**Review by vincentpierre — APPROVED** [review]
Need to edit the documentation before merging.

**Review by hvpeteet — APPROVED** [review]
Looks good to me, we don't use these fields for anything cloud specific.

File: ml-agents/mlagents/trainers/settings.py (hunk 3 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/settings.py:221]: Negative. Updated.

File: ml-agents/mlagents/trainers/settings.py (hunk 5 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/settings.py:186]: I think I'd prefer to use our defaults since it's possible the policy has significantly more capacity than is needed i.e. the Crawler policy of 3/512 vs what we use for the discriminator 2/128. That being said, I also realize this enables users to specify `memory` which we probably want to explicitly prevent in the reward providers. cc @ervteng

Comment (ervteng) [reply ml-agents/mlagents/trainers/settings.py:186]: Not opposed to either route, they have their own pros/cons. Either way as long as it's documented it should be fine. 
Is getting the Policy settings super ugly?

**Review by ervteng — APPROVED** [review]

File: ml-agents/mlagents/trainers/settings.py (hunk 5 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/settings.py:186]: Im not sure how future proof it is for multi-agent scenarios. We could have different policies to select from. Additionally, we currently create reward signals in the `optimizer/torch_optimizer.py` and in the future i think it will be necessary to remove the policy from the optimizer (also for multiagent) in which case this would need to be addressed by either keeping the policy around/moving the creation of the reward provider. My vote is for default network settings
